### PR TITLE
Write packages and vulnerabilities at the time of import

### PR DESCRIPTION
**Pull request by TG1999**
public instance of VCIO contains more than 35 million advisories ( the volume is so high, due to minor changes in every advisory maybe sometime or the changes in models), which takes a lot of time. If we improve the advisories just after the import this will drop volume hugely.

### CODE REVIEW

File: vulnerabilities/improve_runner.py [hunk 1]
```diff
@@ -36,11 +37,18 @@ class ImproveRunner:
     improver and parsing the returned Inferences into proper database fields
     """
 
-    def __init__(self, improver_class):
-        self.improver_class = improver_class
+    def __init__(self, improver_class=None, improver: Improver = None):
```
Comment (pombredanne) [inline vulnerabilities/improve_runner.py:40]: May be keeping the class design and passing kwargs can work too?
```suggestion
    def __init__(self, improver_class, **kwargs):
    self.improver = improver_class(**kwargs)
```

File: vulnerabilities/improve_runner.py [hunk 2]
```diff
@@ -223,10 +231,10 @@ def get_or_create_vulnerability_and_aliases(vulnerability_id, alias_names, summa
         vulnerability.save()
 
     if summary and summary != vulnerability.summary:
-        logger.warn(
-            f"Inconsistent summary for {vulnerability!r}. "
-            f"Existing: {vulnerability.summary}, provided: {summary}"
-        )
+        summary = summary.strip()
```
Comment (pombredanne) [inline vulnerabilities/improve_runner.py:234]: This may be best in another PR

**Review by pombredanne — CHANGES_REQUESTED** [review]
Some nits for your consideration.

**Comment by TG1999** [comment]
@DennisClark we need your input on this, we use advisories as a screenshot of the status of the advisory at a given point of time. A single advisory goes through many improver processes to draw inferences, what should be the best possible way to get a status on advisory model to know which improvers have already ran through a particular advisory ?

**Comment by DennisClark** [comment]
@TG1999   Here are a few thoughts about advisories: 
* every advisory should be associated with a "date_detected" (or similar), which can be the date+time that the advisory was originally posted to VC. 
* since the number and types of potential improvers can vary, we may want to create a field (perhaps "associated_improvers") that can contain one or more pairs of improver_name:improver_date values, perhaps in a json structure or similar, that is updated whenever an improver is processed on the advisory 
* alternatively, the new field would initially identify all the potential improver_name values with a null in the improver_date that is set when the improver is actually run for that advisory
* all of those new elements should be visible in the VCIO UI. 
* the process(es) to initiate advisory improvements should select those where an appropriate improver has not yet been processed for that advisory. 

I don't think we necessarily need a "status" on the advisory, since the list of improver_name:improver_date values ought to give us what would be really useful here (what has been run and when).

**Review by pombredanne — CHANGES_REQUESTED** [review]
Here are some nits for your consideration

File: vulnerabilities/import_runner.py [hunk 3]
```diff
@@ -41,42 +55,257 @@ def run(self) -> None:
...
-            logger.debug(f"Advisory with aliases: {obj.aliases!r} already exists. Skipped.")
 
-    return count
+            if not reference:
+                reference = create_valid_vulnerability_reference(
+                    reference_id=ref.reference_id,
+                    url=ref.url,
+                )
+                if not reference:
+                    continue
```
[the remark this answers is not in the document]
Comment (TG1999) [reply vulnerabilities/import_runner.py:171]: We are logging it already here https://github.com/nexB/vulnerablecode/pull/1280/files#diff-11856d929972b1aa71dcd476298a831d6b598b291c0638057af6d62050ba4a15R225

File: vulnerabilities/import_runner.py [hunk 4]
```diff
@@ -41,42 +55,257 @@ def run(self) -> None:
...
-                affected_packages=[pkg.to_dict() for pkg in data.affected_packages],
-                references=[ref.to_dict() for ref in data.references],
-                date_published=data.date_published,
-                weaknesses=data.weaknesses,
-                defaults={
-                    "created_by": importer_name,
-                    "date_collected": datetime.datetime.now(tz=datetime.timezone.utc),
-                },
-            )
+            self.do_import(advisories)
```
[the remark this answers is not in the document]
Comment (TG1999) [reply vulnerabilities/import_runner.py:119]: Opened follow-up issue here https://github.com/nexB/vulnerablecode/issues/1300

File: vulnerabilities/import_runner.py [hunk 5]
```diff
@@ -41,42 +55,257 @@ def run(self) -> None:
         importer_class = self.importer_class
         logger.info(f"Starting import for {importer_name}")
         advisory_datas = importer_class().advisory_data()
-        count = process_advisories(advisory_datas=advisory_datas, importer_name=importer_name)
+        count = self.process_advisories(advisory_datas=advisory_datas, importer_name=importer_name)
         logger.info(f"Finished import for {importer_name}. Imported {count} advisories.")
 
+    def do_import(self, advisories) -> None:
+        improver = AdvisoryBasedDefaultImprover(advisories=advisories)
```
[the remark this answers is not in the document]
Comment (TG1999) [reply vulnerabilities/import_runner.py:62]: Done!

File: vulnerabilities/migrations/0040_remove_advisory_date_improved_advisory_date_imported.py [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+class Migration(migrations.Migration):
+
+    dependencies = [
+        ("vulnerabilities", "0039_alter_vulnerabilityseverity_scoring_system"),
+    ]
+
+    operations = [
+        migrations.RemoveField(
+            model_name="advisory",
+            name="date_improved",
```
[the remark this answers is not in the document]
Comment (TG1999) [reply vulnerabilities/migrations/0040_remove_advisory_date_improved_advisory_date_imported.py:15]: We have replaced `date_improved` with `date_imported` here, `date_improved` used to be updated every time when any improver runs on the advisory, which was not helping

File: vulnerabilities/import_runner.py [hunk 7]
```diff
@@ -41,42 +55,259 @@ def run(self) -> None:
...
+                    advisory_data=advisory.to_advisory_data()
+                )
+                process_inferences(
+                    inferences=inferences,
+                    advisory=advisory,
+                    improver_name=improver_name,
+                )
+            except Exception as e:
+                logger.info(f"Failed to process advisory: {advisory!r} with error {e!r}")
+        logger.info("Finished improving using %s.", advisory_importer.__class__.qualified_name)
```
Comment (pombredanne) [inline vulnerabilities/import_runner.py:81]:
```suggestion
        logger.info("Finished importing using %s.", advisory_importer.__class__.qualified_name)
```

**Review by pombredanne — COMMENTED** [review]
Here are a few nits... Either apply now or later!
LGTM otherwise!

**Pull request merged**